sw/host/ot_certs/src/x509.rs

```diff
@@ -0,0 +1,255 @@
+                generator.add_variable(s, target_type, &value.s);
+            } else {
+                generator.add_hidden_variable("__hidden_sig_ec_r".to_string(), r);
+                generator.add_hidden_variable("__hidden_sig_ec_s".to_string(), s);
+            }
+            Ok(())
+        }
+    }
+}
+
+/// Generate an X509 certificate using a template certificate and an assignment
+/// of the variables to their specific values.
+pub fn generate_certificate(
+    tmpl: &template::Template,
+    clear_fields: bool,
+) -> Result<CertificateWithVariables> {
+    let mut generator = OffsetGenerator::new(&tmpl.variables);
+
+    let mut builder = X509::builder()?;
+
+    // The certificate expiration time is fixed by the DICE
+    // specification to be "never". The "not before" date is
+    // chosen to be in the past since the device does not
+    // have a reliable way to get the time.
+    let not_before = Asn1Time::from_str("20230101000000Z")?;
+    builder.set_not_before(&not_before)?;
+    let not_after = Asn1Time::from_str("99991231235959Z")?;
+    builder.set_not_after(&not_after)?;
+
+    // Fixed by the DICE specification.
+    builder.set_version(3)?;
+
+    // Serial number.
+    let serial_number = generator
+        .get_value(&tmpl.certificate.serial_number)
+        .expect("cannot get value for the certificate serial number");
+    let serial_number_asn1 = Asn1Integer::from_bn(&bignum_from_bigint(&serial_number))?;
+    builder.set_serial_number(&serial_number_asn1)?;
+
+    // Issuer.
+    let mut issuer_name_builder = X509NameBuilder::new()?;
+    for (key, value) in &tmpl.certificate.issuer {
+        let value = generator.get_value(value).with_context(|| {
+            format!(
+                "cannot get value of key '{}' in the certificate issuer",
+                key
+            )
+        })?;
+        if *key == template::AttributeType::Country && value.len() > 2 {
+            bail!("The country name must fit on two characters");
+        }
+        issuer_name_builder.append_entry_by_nid(key.nid(), &value)?;
+    }
+    let issuer_name = issuer_name_builder.build();
+    builder.set_issuer_name(&issuer_name)?;
+
+    // Subject.
+    let mut subject_builder = X509NameBuilder::new()?;
+    for (key, value) in &tmpl.certificate.subject {
+        let value = generator.get_value(value).with_context(|| {
+            format!(
+                "cannot get value of key '{}' in the certificate subject",
+                key
+            )
+        })?;
+        if *key == template::AttributeType::Country && value.len() > 2 {
+            bail!("The country name must fit on two characters");
+        }
+        subject_builder.append_entry_by_nid(key.nid(), &value)?;
+    }
+    let subject = subject_builder.build();
+    builder.set_subject_name(&subject)?;
+
+    // Random public key.
+    let pubkey = get_pubkey(&mut generator, &tmpl.certificate.subject_public_key_info)?;
```

Comment:
This is not a random key is it?

Reply:
Due to the way openssl work, we have to generate a random key (unless it's completely specified in the template with raw values for x/y), then use this value to extract the offset and clear the values.

Reply:
ah I see, thanks for clarifying. Perhaps a comment here explaining might help?